### DISCUSSION TITLE
Move Controller outside of configurations data class

### DISCUSSION THREAD

**Discussion by EchoEllet** [singerdmx/flutter-quill#2037] [Feature Request]
For now both `QuillEditorConfigurations` and `QuillSimpleToolbarConfigurations` require the `QuillController` which is mutable and requires to removal of the `const` keyword when creating an instance of the configuration class

```dart
QuillEditor.basic(
  configurations: const QuillEditorConfigurations(
    // Invalid constant value.dart(invalid_constant)
    controller: _controller,
  ),
)
```

<details>
<summary>Screenshot</summary>

![Invalid constant value.dart(invalid_constant)](https://github.com/user-attachments/assets/91f694c3-c2e2-4475-8cf4-9e935317962c)

</details>

The `QuillController` is required regardless of the configurations.

Moving the `controller` property outside of `QuillEditorConfigurations` and `QuillSimpleToolbarConfigurations`
will make it possible to use the `const` keyword or not have to create an instance of the configuration class. 

```dart
QuillEditor.basic(
  controller: _controller,
  configurations: const QuillEditorConfigurations(
    // ...
  ),
)
```

If there is a value passed to a property in `QuillEditorConfigurations` that needs to be computed in runtime then the `const` will need to be removed as well though for most use cases, the developer doesn't need to.

We could do this without breaking change by making both not required and temporarily using `assert` to check if one of them is null, deprecate the old one and once we remove it we no longer need to check it using `assert`.

**Comment by AtlasAutocode** [comment]
The configurations seem to be used as a convenient way to set and access the QuillController which is required for all operations.

I did a quick search for 'configurations.controller' and found 17 usages. These seem to be confined to editor and toolbar, so if these were supplied with the controller directly, you would not need access via the configurations.

Developers will need to move setting the controller from configurations to whatever since the same controller instance must be available to the toolbar and editor. Is this a breaking change?

Would it be possible to link toolbar and editor such that the controller only needs to be set in one place? If this is possible then most developers would not even need to set a controller - a default plain controller would suffice.

Other than that, I totally agree with you - it makes no sense to have the controller in the configurations. Configurations should be const.

**Comment by AtlasAutocode** [comment]
@EchoEllet Are you working on this?

I have a need to access some of the configurations from the controller and would like to move the controller out of the configurations as you are recommending.

It would be good to eventually have all of the configurations as const's and sourced from a single common point.

If you are not already working on this, I can start the re-organization.

I would propose either a 1-step breaking change or a 2-step with intermediate non-breaking change.

Since you are more experienced in this project, I defer to your better judgement or I can outline my thoughts.

**Reply by EchoEllet** [reply]
> Are you working on this?

I haven't done much since 2024 as I'm no longer active maintainer for now.

> Since you are more experienced in this project, I defer to your better judgement or I can outline my thoughts.

Depends on the context. For most parts you're working on, you're more familiar than I'm. 

> It would be good to eventually have all of the configurations as const's and sourced from a single common point

We also could use code generation to automate the process or use the new Dart macros. Take a look at [data_class](https://pub.dev/packages/data_class) and the new pre-release version of `freezed` or [Dart macros](https://dart.dev/language/macros). Currently, this feature is in the experimental phase.

> If you are not already working on this, I can start the re-organization.
> I would propose either a 1-step breaking change or a 2-step with intermediate non-breaking change.

In general, I appreciate your work regardless if you do.

The internal docs of the project could use some improvements. For now, some essential parts of the project might not be quite documented to new contributors. For example, the main difference between`Delta` and `Document`

This is something we plan on improving.

**Reply by CatHood0** [reply]
> the main difference between `Delta` and `Document`

This is a good idea. I thinking on improve some docs (like `Delta introduction`) to be more easy to understand and add more docs (like for customize the behavior of `CopyCutServiceProvider` or `Custom Embed Blocks` that could not be clear).

**Reply by AtlasAutocode** [reply]
Thanks for your encouragement.
I will start cautiously with my current fix - Copy-Paste of Multiple line selection with attributes and embeds is broke.
Then proceed to start a non-breaking reorganization of the configurations - will look at your referrals.

**Comment by AtlasAutocode** [comment]
I've submitted a PR to start the process.
First step has enabled controller to be set directly in the editor and toolbar. This is non-breaking since there is a failsafe to get the controller from the configurations. I've also marked the configuration's controller parameter as deprecated and have tried to update the documentation to encourage users to make the changes needed.

At some point, I will remove the deprecated controller parameter in the configurations and that will enable configurations to be const. This will be a breaking change.

@EchoEllet When do you recommend making the breaking change? Wait 2 or 4 weeks in case anyone reports a problem?

**Reply by EchoEllet** [reply]
> When do you recommend making the breaking change? Wait 2 or 4 weeks in case anyone reports a problem?

I would recommend waiting for a bit longer and planning all breaking changes at once. For example, we already have a TODO about a breaking change in `flutter_quill_extensions` to avoid exposing all API as public.

Most of them are minor and require minimal changes.

**Reply by AtlasAutocode** [reply]
OK, will you give me a heads-up when you are about to release the breaking change so we can coordinate?
I don't expect this one to be difficult (famous last words) but will probably do a dry run to document what will be needed when the time comes.

**Reply by EchoEllet** [reply]
> when you are about to release the breaking change so we can coordinate?

Currently, I can't due to the time constraints.

I would suggest creating an issue that documents the next breaking changes for `11.0.0` and there will be a discussion on what will be removed or changed.

You can only add this change for now and I will try to update the thread once I have the time.
